planner/core/rule_predicate_push_down.go

```diff
@@ -638,3 +652,119 @@ func (p *LogicalMemTable) PredicatePushDown(predicates []expression.Expression)
 func (*ppdSolver) name() string {
 	return "predicate_push_down"
 }
+
+// AddPrefix4ShardIndexes Add expression prefix for shard index. e.g. an index is test.uk(tidb_shard(a), a).
+// It transforms the sql "SELECT * FROM test WHERE a = 10" to
+// "SELECT * FROM test WHERE tidb_shard(a) = val AND a = 10", val is the value of tidb_shard(10).
+// It also transforms the sql "SELECT * FROM test WHERE a IN (10, 20, 30)" to
+// "SELECT * FROM test WHERE tidb_shard(a) = val1 AND a = 10 OR tidb_shard(a) = val2 AND a = 20"
+// @param[in] conds            the original condtion of this datasource
+// @retval - the new condition after adding expression prefix
+func (ds *DataSource) AddPrefix4ShardIndexes(sc sessionctx.Context, conds []expression.Expression) []expression.Expression {
+	if !ds.containExprPrefixUk {
+		return conds
+	}
+
+	var err error
+	newConds := make([]expression.Expression, 0, len(conds))
+	newConds = append(newConds, conds...)
+
+	for _, path := range ds.possibleAccessPaths {
+		if path.IsTablePath() || !path.IsUkShardIndex() {
+			continue
+		}
+		newConds, err = ds.addExprPrefixCond(sc, path, newConds)
+		if err != nil {
+			logutil.BgLogger().Error("Add tidb_shard expression failed", zap.Error(err))
```

Comment:
@you06 @cfzjywxk PTAL again!
If any errors produced, don't add tidb_shard expression and log it.

Reply:
The information may not be enough, we could print more about:
- connection id
- table name
- index name and column